metrics/cmd/heapster-apiserver/app/server.go

```diff
@@ -85,78 +60,35 @@ func NewHeapsterApiServer(s *options.HeapsterRunOptions, metricSink *metricsink.
-		OIDCGroupsClaim:   s.OIDCGroupsClaim,
-		KeystoneURL:       s.KeystoneURL,
-	})
-	if err != nil {
-		glog.Fatalf("Invalid Authentication Config: %v", err)
+	genericAPIServerConfig := genericapiserver.NewConfig()
+	if _, err := genericAPIServerConfig.ApplySecureServingOptions(s.SecureServing); err != nil {
+		return nil, err
 	}
 
-	authorizationModeNames := strings.Split(s.AuthorizationMode, ",")
-	authorizationConfig := genericauthorizer.AuthorizationConfig{
-		PolicyFile:                  s.AuthorizationPolicyFile,
-		WebhookConfigFile:           s.AuthorizationWebhookConfigFile,
-		WebhookCacheAuthorizedTTL:   s.AuthorizationWebhookCacheAuthorizedTTL,
-		WebhookCacheUnauthorizedTTL: s.AuthorizationWebhookCacheUnauthorizedTTL,
-		RBACSuperUser:               s.AuthorizationRBACSuperUser,
-	}
-	authorizer, err := genericauthorizer.NewAuthorizerFromAuthorizationConfig(authorizationModeNames, authorizationConfig)
-	if err != nil {
-		glog.Fatalf("Invalid Authorization Config: %v", err)
+	if !s.DisableAuthForTesting {
+		if _, err := genericAPIServerConfig.ApplyDelegatingAuthenticationOptions(s.Authentication); err != nil {
+			return nil, err
+		}
+		if _, err := genericAPIServerConfig.ApplyDelegatingAuthorizationOptions(s.Authorization); err != nil {
+			return nil, err
+		}
 	}
 
-	admissionControlPluginNames := strings.Split(s.AdmissionControl, ",")
+	var err error
 	privilegedLoopbackToken := uuid.NewRandom().String()
-
-	client, err := s.NewSelfClient(privilegedLoopbackToken)
-	if err != nil {
-		glog.Errorf("Failed to create clientset: %v", err)
+	if genericAPIServerConfig.LoopbackClientConfig, err = genericAPIServerConfig.SecureServingInfo.NewSelfClientConfig(privilegedLoopbackToken); err != nil {
+		return nil, err
 	}
 
-	sharedInformers := informers.NewSharedInformerFactory(client, 10*time.Minute)
-	pluginInitializer := admission.NewPluginInitializer(sharedInformers)
-
-	admissionController, err := admission.NewFromPlugins(client, admissionControlPluginNames, s.AdmissionControlConfigFile, pluginInitializer)
+	genericAPIServerConfig.SwaggerConfig = genericapiserver.DefaultSwaggerConfig()
 
-	genericConfig := genericapiserver.NewConfig(s)
-	// TODO: Move the following to generic api server as well.
-	genericConfig.Authenticator = authn
-	genericConfig.SupportsBasicAuth = len(s.BasicAuthFile) > 0
-	genericConfig.Authorizer = authorizer
-	genericConfig.AdmissionControl = admissionController
-	genericConfig.APIResourceConfigSource = storageFactory.APIResourceConfigSource
-	genericConfig.MasterServiceNamespace = s.MasterServiceNamespace
-	genericConfig.Serializer = api.Codecs
-
-	// TODO: Move this to generic api server (Need to move the command line flag).
-	if s.EnableWatchCache {
-		cachesize.SetWatchCacheSizes(s.WatchCacheSizes)
-	}
+	// TODO: Include EnableWatchCache and other GenericServerRunOptions
+	// TODO: AdvertiseAddress?
 
-	return genericConfig.New()
+	return genericAPIServerConfig.Complete().New()
```

Comment:
got to the end here.  At a glance, this looks good.  Did you support a general heapster webhook authentication before?